adrf/pagination.py

```diff
@@ -0,0 +1,151 @@
+"""
+Pagination serializers determine the structure of the output that should
+be used for paginated responses.
+"""
+from django.core.paginator import InvalidPage
+from rest_framework.exceptions import NotFound
+from rest_framework.pagination import _reverse_ordering, \
+    PageNumberPagination as DRFPageNumberPagination, \
+    LimitOffsetPagination as DRFLimitOffsetPagination, \
+    CursorPagination as DRFCursorPagination
+
+
+class PageNumberPagination(DRFPageNumberPagination):
+    async def paginate_queryset(self, queryset, request, view=None):
+        """
+        Paginate a queryset if required, either returning a
+        page object, or `None` if pagination is not configured for this view.
+        """
+        self.request = request
+        page_size = self.get_page_size(request)
+        if not page_size:
+            return None
+
+        paginator = self.django_paginator_class(queryset, page_size)
+        page_number = self.get_page_number(request, paginator)
+
+        try:
+            self.page = paginator.page(page_number)
+        except InvalidPage as exc:
+            msg = self.invalid_page_message.format(
+                page_number=page_number, message=str(exc)
+            )
+            raise NotFound(msg)
+
+        if paginator.num_pages > 1 and self.template is not None:
+            # The browsable API should display pagination controls.
+            self.display_page_controls = True
+
+        return self.page
+
+
+class LimitOffsetPagination(DRFLimitOffsetPagination):
+    async def aget_count(self, queryset):
+        """
+        Determine an object count, supporting either querysets or regular lists.
+        """
+        try:
+            return (await queryset.acount())
+        except (AttributeError, TypeError):
+            return len(queryset)
+
+    async def paginate_queryset(self, queryset, request, view=None):
+        self.request = request
+        self.limit = self.get_limit(request)
+        if self.limit is None:
+            return None
+
+        self.count = await self.aget_count(queryset)
+        self.offset = self.get_offset(request)
+        if self.count > self.limit and self.template is not None:
+            self.display_page_controls = True
+
+        if self.count == 0 or self.offset > self.count:
+            return []
```

Comment:
I've used this implementation of `LimitOffsetPagination` successfully. The only change I needed was to this empty response so you can always use `async for` on it.
```suggestion
            return queryset.none()
```